delegation command for notary-cli

riyazdf opened the pull request:
Adds a `notary delegation` command to the cli.  There are three commands:
- `add`: adds a delegation using a given gun, X509 cert(s) with public key, role (ex: "targets/delegation"), and optional comma-separated `--paths`. Currently we only validate the expiry date on the certificate since there are no restrictions on the signature, but it could be backed by the TUF root key or existing PKI in the future. This command creates a TUF change offline, which will need to be published to take effect.
- `remove`: removes delegation info using a gun and role with either key IDs (standard arguments) or comma-separated `--paths`.  The command can also just remove all information for the specified role using a `--force` flag (this requires confirmation).  Creates a TUF change offline (update or delete, depending on `--force`) that will require a publish for effect.
- `list`: updates the notary repository online before listing all published delegations in a table format.

Closes #434.

NathanMcCauley commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	// Add the delegation to the repository
+	err = nRepo.AddDelegation(role, 1, []data.PublicKey{pubKey}, paths)
```
What is `1` in this context? Can we make a const with a descriptive name?

riyazdf replied:
This is the number of thresholds, I can definitely make a const to make things clearer.  We could also try to make this configurable, though I don't think we've surfaced a way to add keys to non-root roles or configure role thresholds at the moment

NathanMcCauley commented on trustmanager/x509utils.go:
```diff
@@ -299,6 +299,39 @@ func ParsePEMPrivateKey(pemBytes []byte, passphrase string) (data.PrivateKey, er
+// Currently, this is only a time expiry check
+func ValidateCertificate(c *x509.Certificate) error {
+	now := time.Now()
+	if now.Before(c.NotBefore) || now.After(c.NotAfter) {
```
Sometimes makes sense to have a bit of a grace period on newly created certs, in case it was just created. Maybe give it a day of leeway on creation time?

riyazdf replied:
Good idea, will do

endophage commented on client/client.go:
```diff
@@ -509,6 +509,58 @@ func (r *NotaryRepository) GetChangelist() (changelist.Changelist, error) {
+	}
+
+	// All top level delegations (ex: targets/level1) are stored exclusively in targets.json
+	targetsJSON, err := r.fileStore.GetMeta("targets", 0)
```
After calling `r.Update()`, the validated repository is already de-serialized into `r.tufRepo`. No need to load the files from the filestore.

endophage commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,174 @@
+	keyID := pubKey.ID()
+
+	// Check to see if it is an invalid ID
+	if len(keyID) != idSize {
```
Because we're generating this `keyID` via `pubKey.ID()` it's redundant and inflexible to have this check, should just be deleted.

endophage commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,174 @@
+	paths := args[3:]
+
+	// Check to see if the public key file exists
+	if _, err := os.Stat(pubKeyPath); os.IsNotExist(err) {
```
Can probably just do the `ioutil.ReadFile` below and not bother with the `Stat`.

endophage commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,174 @@
+
+// delegationAdd adds a public key from a certificate to a specific role in a GUN
+func (d *delegationCommander) delegationAdd(cmd *cobra.Command, args []string) error {
+	if len(args) < 4 {
```
Add a short test with 3 args just to his this case?

riyazdf replied:
good idea, done!

endophage commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,174 @@
+
+// delegationsList lists all the delegations for a particular GUN
+func (d *delegationCommander) delegationsList(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
```
Add a short test with an incorrect number of args.

riyazdf replied:
done!  Also for remove too

endophage commented on client/client.go:
```diff
@@ -509,6 +509,58 @@ func (r *NotaryRepository) GetChangelist() (changelist.Changelist, error) {
+		if err != nil {
+			return nil, err
+		}
+		for _, d := range delegationMeta.Signed.Delegations.Roles {
```
replace this whole loop with 

```
allDelegations = append(allDelegations, delegationMeta.Signed.Delegations.Roles...)
delegationsList = append(delegationsList, delegationMeta.Signed.Delegations.Roles...)
```

riyazdf replied:
go is awesome, thanks for the tip!

endophage replied:
Yeah, it's nice :-) 

I actually saw this and knew I'd done something similar elsewhere so I had to go double check my code. I was doing map->slice conversion so unfortunately couldn't make use of this.

endophage commented on client/client.go:
```diff
@@ -509,6 +509,51 @@ func (r *NotaryRepository) GetChangelist() (changelist.Changelist, error) {
+// GetDelegationRoles returns the keys and roles of the repository's delegations
+func (r *NotaryRepository) GetDelegationRoles() ([]*data.Role, error) {
+	// Update state of the repo to latest
+	_, err := r.Update(false)
```
As we're not using the non-err return value:
```
if _, err := r.Update(false); err != nil {
    return nil, err
}
```

endophage replied:
Also, for test purposes, if I remember right you can make `Update` fail by setting `r.tufRepo = nil`

endophage commented on client/client.go:
```diff
@@ -509,6 +509,51 @@ func (r *NotaryRepository) GetChangelist() (changelist.Changelist, error) {
+		delegationsList = delegationsList[1:]
+
+		// Get metadata
+		delegationMetaJSON, err := r.fileStore.GetMeta(delegation.Name, 0)
```
Can replace this like you did above:
```
delegationMeta, ok := r.tufRepo.Targets[delegation.Name]
```

endophage commented on cmd/notary/delegations.go:
```diff
@@ -0,0 +1,164 @@
+	// initialize repo with transport to get latest state of the world before listing delegations
+	nRepo, err := notaryclient.NewNotaryRepository(config.GetString("trust_dir"), gun, getRemoteTrustServer(config), getTransport(config, gun, true), retriever)
+	if err != nil {
+		fatalf(err.Error())
```
There's a mix of `fatalf` and `return err`, it seems like it should always be `return err` based on function signature?

riyazdf commented:
notes on my end after some testing and discussion this evening, marking as WIP and open to discussion:

- I'm going to change how `notary delegation add` behaves from this PR's implementation, so that we add new keys if possible (this PR overwrites).  If the `paths` conflict between the existing delegation and the one we want to add with the same name, we reject the new delegation -- this check must be done online to avoid race conditions.

- I'll also change the `notary delegation remove` command to remove key IDs one by one, and remove the delegation once no keys remain.  This will probably change `RemoveDelegation` as well, and also has to be checked online for race conditions (if another client concurrently adds, etc.)

endophage commented:
Been thinking for consistency with the other sub-commands, which are plural nouns (`keys` and `certs`), we might also want to make it `notary delegates ...`

riyazdf commented:
@endophage unless I'm missing something, I think all of notary's sub-commands are singular, so maybe leave as `notary delegation ...` or `notary delegate ...`?

endophage commented:
Wow... yeah, I had a complete mental inversion. I prefer `notary delegate`. Happy to bow to the consensus though if people disagree, it's only a 2 character difference.

riyazdf commented:
removed WIP -- I had to change the logic in `helpers.go` to handle adding and removing one key at a time from a delegation:
- If the delegation role already exists, `delegation add` will add to the pre-existing delegation with the new `keyID` if and only if the paths match.  If paths do not match, this change is rejected on publish.
- `delegation remove` will remove single keys by `keyID` at a time until the last key, when the delegation is deleted in entirety

endophage commented:
When adding a key to an existing delegation, having to duplicate the paths seems like terrible UX. Would it be possible to add a `--update` flag to the add command that indicates that the key and any new paths should be added to the role, regardless of keys or paths it currently contains?

riyazdf commented:
updating this PR for documentation, after some offline discussion with @endophage, it seems most intuitive to eventually have finer grained flags such as `--key` and `--path` to operate on these single items within a `delegation add` or `delegation remove` command.

riyazdf commented:
Removing WIP.

I added the ability for multiple keys/paths to add/remove.  Keys are the default argument, paths are specified by a flag `--paths` which takes a comma-separated list of paths to change: `notary delegation add GUN ROLE KEY1 KEY2 --paths=PATH1,PATH2`

I also added a force flag to remove, so an admin can easily `notary delegation remove GUN ROLE --force` to remove all data (there's an accompanying `--yes` flag to avoid a confirmation)

endophage commented:
Do we need `--force` and `--yes`? What's the difference in behaviour between `notary delegation remove GUN ROLE` and `notary delegation remove GUN ROLE --force`?

riyazdf commented:
I could go either way with having the `--force` flag but I think we should at least prompt for confirmation ( and have a `--yes` flag to bypass this like when we remove certs) if you're about to delete all information for a delegation role.

Right now `notary delegation remove GUN ROLE` is an invalid command unless `--force` is included.

diogomonica commented:
Haven't tried this since last week, but code LGTM.

diogomonica commented:
@riyazdf rebase please

riyazdf commented:
@diogomonica done!

diogomonica commented:
@riyazdf me and @endophage have a long standing disagreement regarding whether notary should be 100% invocable from the cmd-line without expect scripts (the --yes was added by me a while back).

I think we need @cyli to weigh in and decide.

cyli commented:
@diogomonica what is the agreement?  Whether `--yes` should be included?

Maybe it would be less confusing if `--force` is changed to `--all` or `--delegation` for removing all delegation information?  `--force` to me already implies that it will just proceed without asking.  Aside from that, I'm mostly indifferent between `--force` and `--yes`.

diogomonica commented:
I like --yes for stuff that would ask and won't. I agree with --all.

riyazdf commented:
I also agree with `--all` (just updated)

endophage commented:
@diogomonica I'm fine with having one of `--yes` and `--force`, I was just questioning that they appeared to be redundant.

I don't think `--all` semantically makes sense. If the command is `notary delegation remove GUN DELEGATION`, what does `--all` mean? The `DELEGATION` is singular. I would say we should just have `notary delegation remove GUN DELEGATION` work with a prompt, and add the `--yes` flag to preemptively confirm (i.e. no prompt).

diogomonica commented:
Agreed with `notary delegation remove GUN DELEGATION` + `--yes`

riyazdf commented:
I think this is good, I was originally thinking that `--all` made sense to specify multiple parts of a delegation but I think keeping abstracting it to a singular delegation makes sense.  Just updated.

cyli commented on cmd/notary/integration_test.go:
```diff
@@ -144,6 +145,223 @@ func TestClientTufInteraction(t *testing.T) {
 }
 
+// Initialize repo and test delegations commands by adding, listing, and removing delegations
+func TestClientDelegationsInteraction(t *testing.T) {
```
This is an awesome integration test!

cyli commented:
LGTM!
